### PR TITLE
test(llama): fix the long-standing GQA backward failure (pin SDPA backend)

### PR DESCRIPTION
**Pull request by saforem2**
## The failure

`test_gqa_matches_repeat_kv_backward` has failed **deterministically** on macOS/CPU since it landed in #199. It's the reason every test report in recent PRs has read "1 failed" and been waved off as pre-existing:

```
Mismatched elements: 3 / 1024 (0.3%)
Greatest absolute difference: 1.8849968910217285e-06 (up to 1e-06 allowed)
Greatest relative difference: 0.00015726150013506413 (up to 1e-05 allowed)
```

## Root cause: unpinned SDPA backend

The test doesn't pin a backend, so it runs under whichever torch picks. Measured on this config (torch 2.12.1, CPU fp32) — max `|grad diff|` between the `enable_gqa=True` and materialized `repeat_kv` paths:

| backend | max abs grad diff |
|---|---|
| **MATH** | **0.000e+00** (bit-identical) |
| **FLASH** | **7.629e-06** (exceeds `atol=1e-6`) |

FLASH fuses attention and reassociates the backward reduction differently for broadcast kv vs materialized kv. That's floating-point reassociation, **not** a math difference — which is exactly what the test claims to pin. So the test is right about the property and wrong about the conditions.

This also explains the platform split: CI (Linux) selected MATH and passed; macOS selected FLASH and failed.

## Fix

- **Pin the strict test to `SDPBackend.MATH`**, where `atol=1e-6` means what it says and the comparison is exact.
- **Add `test_gqa_matches_repeat_kv_backward_flash`** at an fp32-appropriate tolerance (`rtol=2e-4`, `atol=1e-4`), skipped when the backend is unavailable.

I deliberately did **not** just loosen the original tolerance. That would have hidden which backend was drifting and by how much, and silently weakened the exact-equivalence check that is the whole point of the test. This keeps the exact assertion **and** adds coverage of the backend that actually runs in production.

## Result

```
1386 passed, 11 skipped, 0 failed
```

The suite is fully green — no more "1 pre-existing failure" caveat on every PR.

## Summary by Sourcery

Ensure GQA backward gradients tests are robust across SDPA backends by pinning the strict check to MATH and adding FLASH coverage at appropriate tolerances.

Tests:
- Pin the GQA backward gradient equivalence test to the SDPA MATH backend for bit-identical comparison.
- Add a FLASH-attention GQA backward gradient equivalence test with relaxed fp32-appropriate tolerances and backend-availability guard.

### CODE REVIEW

File: tests/test_llama_attention_gqa.py [hunk 1]
```diff
@@ -29,6 +30,21 @@ def _build_attention(n_heads: int, n_kv_heads: int, head_dim: int = 16):
...
+    k = torch.randn(2, 1, 8, 16)
+    v = torch.randn(2, 1, 8, 16)
+    try:
+        with sdpa_kernel(backend):
+            torch.nn.functional.scaled_dot_product_attention(
+                q, k, v, enable_gqa=True
+            )
+    except Exception:
+        return False
+    return True
```
[the remark this answers is not in the document]
Comment (saforem2) [reply tests/test_llama_attention_gqa.py]: Both points are right — fixed.

The probe now runs **forward + backward** with `requires_grad` tensors and asserts a grad actually materialized. That matters because the gated test compares gradients, so a forward-only probe could report a backend available and then fail inside the test body instead of skipping.

And the catch is narrowed to `(RuntimeError, NotImplementedError)` — how torch signals "this backend cannot run that." Anything else (a `TypeError` from an API signature change, say) now surfaces loudly instead of masquerading as a skip.

Verified FLASH is still detected and `test_gqa_matches_repeat_kv_backward_flash` still **runs** — a stricter probe that silently skipped would have quietly removed the coverage this PR exists to add.
